core/src/main/java/io/seata/core/rpc/netty/RmNettyRemotingClient.java

```diff
@@ -173,42 +167,49 @@ public void onRegisterMsgFail(String serverAddress, Channel channel, Object resp
         RegisterRMResponse registerRMResponse = (RegisterRMResponse)response;
         String errMsg = String.format(
-            "register RM failed. client version: %s,server version: %s, errorMsg: %s, " + "channel: %s", registerRMRequest.getVersion(), registerRMResponse.getVersion(), registerRMResponse.getMsg(), channel);
+                "register RM failed. client version: %s,server version: %s, errorMsg: %s, " + "channel: %s", registerRMRequest.getVersion(), registerRMResponse.getVersion(), registerRMResponse.getMsg(), channel);
         throw new FrameworkException(errMsg);
     }
 
     /**
      * Register new db key.
      *
-     * @param resourceGroupId the resource group id
-     * @param resourceId      the db key
+     * @param resourceId db key
      */
-    public void registerResource(String resourceGroupId, String resourceId) {
-
+    public void registerResource(String resourceId, BranchType type) {
         // Resource registration cannot be performed until the RM client is initialized
         if (StringUtils.isBlank(transactionServiceGroup)) {
             return;
         }
 
-        if (getClientChannelManager().getChannels().isEmpty()) {
-            getClientChannelManager().reconnect(transactionServiceGroup);
-            return;
-        }
-        synchronized (getClientChannelManager().getChannels()) {
-            for (Map.Entry<String, Channel> entry : getClientChannelManager().getChannels().entrySet()) {
-                String serverAddress = entry.getKey();
-                Channel rmChannel = entry.getValue();
-                if (LOGGER.isInfoEnabled()) {
-                    LOGGER.info("will register resourceId:{}", resourceId);
-                }
-                sendRegisterMessage(serverAddress, rmChannel, resourceId);
-            }
+        NettyClientChannelManager channelManager = getClientChannelManager();
+        if (channelManager.getChannels().isEmpty()) {
+            channelManager.reconnect(transactionServiceGroup);
         }
+
+        RegisterRMRequest message = buildRegisterRMRequest(resourceId, type);
+        Map<String, Channel> channels = channelManager.getChannels();
+        channels.forEach((serverAddress, rmChannel) -> {
+            LOGGER.info("will register resource: {}", resourceId);
+            sendRegisterMessage(serverAddress, rmChannel, message);
+        });
     }
 
-    public void sendRegisterMessage(String serverAddress, Channel channel, String resourceId) {
+    /**
+     * build RegisterRMRequest
+     * note: BranchType store in message.extraData
+     */
+    private RegisterRMRequest buildRegisterRMRequest(String resourceId, BranchType type) {
         RegisterRMRequest message = new RegisterRMRequest(applicationId, transactionServiceGroup);
         message.setResourceIds(resourceId);
+        if (type != null) {
+            String typeName = type.name();
+            message.setExtraData(typeName);
```

Comment:
If you need to add additional fields to ExtraData, consider the extensibility design.